In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [4]:
df = pd.read_csv('heart.csv')

In [5]:
df.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [6]:
X = df.drop('target', axis=1)
y = df['target']

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)


In [8]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [9]:
model = LogisticRegression()

In [10]:
model.fit(X_train_scaled, y_train)

# y = w1x1 + w2x2 + w3x3 + b =>  Output => Sigmoid function => Output  => (0-1) => Decision boundary => 0/1


LogisticRegression()

In [11]:
y_pred = model.predict(X_test_scaled)
y_pred

array([1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0,
       1, 1, 1, 1, 1, 0, 0])

In [12]:
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)

0.7951219512195122


In [13]:
#Acutal probabilities calculated by Logisctic Regression

model.predict_proba(X_test_scaled)

array([[2.50899962e-02, 9.74910004e-01],
       [3.08852096e-03, 9.96911479e-01],
       [9.85306183e-01, 1.46938167e-02],
       [2.89761222e-02, 9.71023878e-01],
       [9.57254867e-01, 4.27451327e-02],
       [3.23911438e-01, 6.76088562e-01],
       [9.92688795e-01, 7.31120516e-03],
       [9.95631806e-01, 4.36819389e-03],
       [1.59101721e-01, 8.40898279e-01],
       [9.84720009e-01, 1.52799914e-02],
       [2.02259817e-02, 9.79774018e-01],
       [9.98270123e-01, 1.72987666e-03],
       [1.22076366e-01, 8.77923634e-01],
       [1.31501218e-01, 8.68498782e-01],
       [9.30091858e-01, 6.99081417e-02],
       [5.58945481e-02, 9.44105452e-01],
       [9.93149896e-01, 6.85010394e-03],
       [2.21624400e-02, 9.77837560e-01],
       [2.42784610e-01, 7.57215390e-01],
       [9.88597120e-01, 1.14028800e-02],
       [1.75841778e-01, 8.24158222e-01],
       [9.49727672e-01, 5.02723277e-02],
       [5.09190610e-01, 4.90809390e-01],
       [9.97449408e-01, 2.55059211e-03],
       [1.758417

When we applied linear regression to heart dataset output was

1.7
-0.4
0.82

does this mean negative disease or 1.7 disease

To solve this problem we will use Logistic regression.

Logistic regression uses Sigmoid function to keep the predictions between 0 - 1 (0% to 100%)

Sigmoid function takes any number as input

- 1000
- -500
- 9
- 0

and compresses it into a value between 0 and 1

Logistic regression does not directly predict the categories. it predicts the probabilities instead

0.90		1
0.20 		0

Usually probability >= 0.5 output is 1
probability < 0.5 output is 0

Cutoff point by default is 0.5
this cutoff point is called Decision Boundary

Accuracy measures how many predictions were correct.

90 correct out of 100
Accuracy = 90% or 0.90

[0.0250899962, 0.974910004] = 1

Logistic Regression is used everywhere:

medical diagnosis
fraud detection
spam filtering
loan approval system
customer churn prediction
